### PR TITLE
Rewrite rewrite

### PR DESCRIPTION
**Pull request by atomcult**
An assortment of fixes and an initial rewrite of auto-mount

### CODE REVIEW

**Comment by dilyn-corner** [comment]
This looks quite good and I appreciate the effort!

We'll have to test but I don't see anything obviously wrong. Thanks for also jumping ahead and fixing the things I was going to do after the initial rewrite!

**Comment by atomcult** [comment]
No problem!

Tacked on a bit more that replaces all the code that uses udisks with code that uses inotifywatch, but I haven't yet removed any of the dependencies from the snapcraft.yaml.

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount]: preemptive `rm_mount not found`

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount:23]: We should have a third case, even if it's just `*) ;;`

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount]: Could just `readonly _date=$(date -ISeconds)`, correct?

Comment (atomcult) [reply scripts/bin/auto-mount]: I just do what shellcheck tells me to do
https://www.shellcheck.net/wiki/SC2155

File: scripts/bin/auto-mount
Comment (atomcult) [reply scripts/bin/auto-mount:23]: That's a good point. In theory, that should exhaust all the options, since we're only listening for `moved_to` and `delete` events, but who knows what other garbage could get thrown our way, especially if there's an error

Comment (dilyn-corner) [reply scripts/bin/auto-mount:23]: If we get an error, we just `panic() { panic | panic }`

File: scripts/bin/auto-mount
Comment (dilyn-corner) [reply scripts/bin/auto-mount]: I like intentionally obfuscating return values though :(

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount]: So `>>` instead of `>` would mean that we support installing snaps from more than just one partition. It's not wrong, but I'm not certain that's what we want?

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount]: I think if we're going to have a `common` and we're putting at least one variable in it, there isn't a strong reason against putting every variable in it (at least the _common_ ones).

File: scripts/bin/common
Comment (dilyn-corner) [inline scripts/bin/common]: We don't need a shebang (or `+x`) if it's just being sourced.

File: scripts/bin/auto-mount
Comment (atomcult) [reply scripts/bin/auto-mount]: I'm actually not sure there's a difference with a FIFO. When I originally wrote this, I wasn't thinking about the fact that writing to a FIFO blocks until read from, so I wanted to make sure that each line was appended. I think it's completely reasonable to switch this over to `>`.

Anyhow, try_mount() is always called with a single device as an argument, and every device that's newly populated in `/dev/disk/by-uuid` will be processed, so by default we are actually checking every partition.

Comment (dilyn-corner) [reply scripts/bin/auto-mount]: _it's a fifo_ 
Then this is fine and it's overwhelming pedantry.

File: scripts/bin/common
Comment (atomcult) [reply scripts/bin/common]: I used a shebang rather than have vim magic at the top to set the filetype, since it's more universal.

Comment (dilyn-corner) [reply scripts/bin/common]: What do I do when a person exhibits undefined behavior

File: scripts/bin/auto-mount
Comment (atomcult) [reply scripts/bin/auto-mount:23]: The "If I can't run, nobody can" method of error handling. I like it

File: scripts/bin/auto-mount
Comment (dilyn-corner) [reply scripts/bin/auto-mount]: Note from conversation: unmounting is hard to do in a clean fashion (no good way of indicating a "finished" state). What we can instead do it is unmount the partition and trigger a system reboot, which is a fine enough indication that the process has finished.

We can leave a comment indicating other options one can take, but we shouldn't implement them.

Comment (atomcult) [reply scripts/bin/auto-mount]: We should think about what to do in cases where no snaps are imported. Unfortunately, the best option is probably to simply avoid rebooting in that case, otherwise we get reboots even if we're just trying to import user assertions.

Comment (atomcult) [reply scripts/bin/auto-mount]: Also, this could potentially be a DoS attack, if someone just leaves in a USB drive 😂

File: scripts/bin/auto-install
Comment (dilyn-corner) [inline scripts/bin/auto-install]: I think it's 17.03 minutes

I don't think there's actually a way to fix this; we *have* to communicate with the store and get a response for this to succeed, and there's no way of making that faster or cleaner afaik

File: scripts/bin/auto-install
Comment (dilyn-corner) [inline scripts/bin/auto-install]: Should be `ls "$_mount_point"/*.assert`?
Maybe this should be a `find` instead of an `ls`; supporting arbitrarily nested asserts and snaps is trivial if we use `basename`.

**Comment by dilyn-corner** [comment]
I can't assign anyone here, but @mboudr35 your thoughts are appreciated :)

File: scripts/bin/auto-install
Comment (atomcult) [reply scripts/bin/auto-install]: Yes lol. You would not believe how many times I have made that mistake.
And yeah, I think using `find` would be a good improvement

File: udisksd.patch
Comment (dilyn-corner) [inline udisksd.patch:1]: For posterity: we don't require this patch anymore because...?

Comment (atomcult) [reply udisksd.patch:1]: Because we're not using udisksd anymore. Snapcraft.yaml still doesn't reflect that, but I figured I'd go ahead and clean up the patch and the udisks script

File: scripts/bin/auto-install
Comment (dilyn-corner) [inline scripts/bin/auto-install:31]: Do these quotes do what you think they do?
I think we can skip the inner "" on the substitution.

Comment (atomcult) [reply scripts/bin/auto-install:31]: I dunno, what do you think I think they do?

File: scripts/bin/auto-install
Comment (dilyn-corner) [inline scripts/bin/auto-install]: Is this actually just busy-waiting for a mount to happen? If the mount point exists but there aren't any found assertions at that mount point, I think we just log and exit (there's nothing to do)?

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount:29]: Odd nit: technically we don't have to do a substitution to strip the lagging `/` because (relevant enough) versions of libc collapse particular repetitions of `/` to the expected result. I believe in all cases, `//` is collapsed to `/`. 

I think you only did this to make the code nicer to look at, and I *doubt* there's any actual problem with doing this, but I felt like showing off.

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount]: I think `rm_mount()` is just trivially `umount $_target && shutdown {-h|-H $message}`, but of course the Store team recently made the required interface for powering off a system super-privileged. So I think we may just scrap the method of notifying...

File: scripts/bin/common
Comment (dilyn-corner) [inline scripts/bin/common]: `echo` -> `touch`?

File: scripts/bin/auto-mount
Comment (atomcult) [reply scripts/bin/auto-mount]: Agreed, although we should `sync`, `umount`, and `eject` in `auto-install`, since it's the script that actually knows when all the snaps are done being installed. Once `auto-install` unmounts the drive, inotifywatch will emit a `DELETE` message for that device, and we could optionally remove the directory that it was mounted in. By default, that's in a temporary directory, so it normally shouldn't matter, but `MOUNT_DIR` *is* overrideable.

File: test
Comment (dilyn-corner) [inline test]: Do we want to strip all trailing `/` or just the last one?

Compare:

```
_dir=foo///
echo ${_dir%/}
echo ${_dir%%"${_dir##*[!/]}"}
```

Not certain if it matters but I also don't know why implement `dirname` in `sh` if we already have `/usr/bin/dirname`.

File: scripts/bin/common
Comment (atomcult) [reply scripts/bin/common:10]: As it's currently used, `mklog()` is called just to generate an appropriate name, and is implicitly created when `log()` is called. That beings said, I do think it's probably worth initializing empty logs, though. An empty log is definitely better than no log.

File: test
Comment (dilyn-corner) [inline test:22]: How dare you put this blank line here

File: scripts/bin/common
Comment (atomcult) [reply scripts/bin/common]: As it's currently used, mklog() is called just to generate an appropriate name, and is implicitly created when log() is called. That beings said, I do think it's probably worth initializing empty logs, though. An empty log is definitely better than no log.

**Review by dilyn-corner — COMMENTED** [review]
Left some extra comments. I think we're very close now!

File: scripts/bin/common
Comment (dilyn-corner) [reply scripts/bin/common]: I think it would be good to create the empty log because what happens if we can't actually write to the filesystem (imagine an SD card with write-lock on).

File: scripts/bin/auto-mount
Comment (dilyn-corner) [reply scripts/bin/auto-mount]: Sure, if you want to be *safe about it* :P 

There's no down-side to cleaning up after ourselves even if it's just in `/tmp`; it's probably the correct move because (as you said) `MOUNT_DIR` could be set somewhere else.

File: scripts/bin/auto-install
Comment (dilyn-corner) [reply scripts/bin/auto-install:31]: I don't think you think they do anything :v 

Quotes are vile... I think in the case of `$(...)` nested quotes are fine though...? Not certain. At the very least they'd be safe to remove, but I guess I don't care.

File: udisksd.patch
Comment (dilyn-corner) [reply udisksd.patch:1]: Just in time for us to officially have a udisks2 snap *again*!

File: scripts/bin/auto-install
Comment (atomcult) [reply scripts/bin/auto-install:31]: Yeah, quotes are treated specially within command substitutions. And I think they *are* necessary, since `${2%%_*}` could contain spaces.

Comment (dilyn-corner) [reply scripts/bin/auto-install:31]: It better not!

ETA: ... Ugh. Yes, yes.

File: scripts/bin/auto-install
Comment (atomcult) [reply scripts/bin/auto-install]: Ah, yeah, we should be copying the log over and doing all the unmount stuff even if there are no asserts. Good catch.

File: test
Comment (atomcult) [reply test:22]: Well, I have to inflate my line counts somehow :P

File: scripts/bin/auto-install
Comment (dilyn-corner) [inline scripts/bin/auto-install]: Remake to `eject_device $WATCH_FILE` with `_disk="$1` and `eject $_disk`, or something like?

Comment (atomcult) [reply scripts/bin/auto-install]: Is your issue with implicitly passed arguments (`$_mount_point`) or reverse engineering the block device path?

I don't really like implicitly passing args, but my instinct was to have `_mount_point="$1"`, which felt even worse. Other options are to use a different arg for (maybe just `_mount`) or we could also make the function run in a subshell:

```sh
eject_device() (
  _mount_point="$1"
  
  # ...
)
```

Comment (dilyn-corner) [reply scripts/bin/auto-install]: My thinking was that because we already have a FIFO which holds the path to the partition we can just reuse that

ETA
I'm just now realizing that I misread this block.

`eject` seems to target the entire block device, not just some partition. So the target should be the resolution of the link at `/dev/disk/by-uuid/foo` without the partition number.

Comment (dilyn-corner) [reply scripts/bin/auto-install]: > [If a device partition is specified, the whole-disk device is used.](https://www.man7.org/linux/man-pages/man1/eject.1.html)

Nevermind!

Though I think because we have a mounting script we could probably do with moving the eject logic to their and having this script call that script to do the work... Added benefit of merely getting to reuse `$MOUNT_POINT`...

Comment (atomcult) [reply scripts/bin/auto-install]: Hmm.. that's a good point. Depending on how complicated this ends up being, we could potentially live dangerously and forego ejecting entirely. As it stands, users have to just guess when the drive has been ejected anyway, unless they implement custom notification functionality.

Comment (dilyn-corner) [reply scripts/bin/auto-install]: I think ejecting is the minimum safety level considering we do write to the disk. I'm thinking that we should have an idea surrounding what cleanup looks like which can be extended in the future (i.e. notifications).

Comment (dilyn-corner) [reply scripts/bin/auto-install]: That being said, I think I'm satisfied by this PR at this point.

File: scripts/bin/auto-install
Comment (dilyn-corner) [inline scripts/bin/auto-install:7]: We never call `mklog` with any args; is this a forward-looking thing and if so, what is the intention?

File: scripts/bin/auto-install
Comment (dilyn-corner) [inline scripts/bin/auto-install:12]: `_log_path` feels like a misnomer as `_log_dir` should already be the path. Should we use `_log_file` instead, with it being just the name and have `_log_path` be `${_log_dir}/${_log_file}`?

File: scripts/bin/auto-install
Comment (dilyn-corner) [inline scripts/bin/auto-install:150]: If we take my above comment about `$_log_file` we can change this to `rm -rf $_log_dir` - this is a reminder note :)

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount]: I don't think we need to watch the 'delete' event if we only have the `MOVED_TO` case.

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount]: Should we be `/dev/null`ing the output of this command? It seems relevant for e.g. debugging.

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount]: I think this should be `rmdir -p "$_target"` to avoid my fun issue of wiping my whole USB stick.

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount:72]: I dislike parsing `ls` but `find` here is... unpleasantly unsupported by POSIX...
I think this stays but we should remember to resolve this; mark with TODO

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount]: Pedantic: `[ -z "$_uuids" ] || { ... }`, but I don't mind.

File: scripts/bin/auto-mount
Comment (dilyn-corner) [inline scripts/bin/auto-mount:78]: Should we use `rmdir -p "${MOUNT_DIR}/${_uuid}"` to ensure the temp dir is cleaned for this case?

Comment (dilyn-corner) [reply scripts/bin/auto-mount:78]: I guess it doesn't matter because we `rmdir "$MOUNT_DIR"` on line 63 :v

**Review by dilyn-corner — COMMENTED** [review]
I think we're done here. I've left some comments but none of them block merging this.

I'm going to go through and leave some TODOs and potentially add some helpful comments to document things. After I've gotten some +1s we can merge :)

**Review by dilyn-corner — APPROVED** [review]

**Pull request merged**
